src/stream/src/executor/gap_fill.rs

```diff
@@ -0,0 +1,1905 @@
+        time_row.memcmp_serialize(&self.time_key_serde)
+    }
+
+    /// Scans a range of data from `StateStore` to build a window of rows before `end_time`.
+    pub async fn scan_range_before(
+        &self,
+        end_time: &GapFillCacheKey,
+        limit: usize,
+    ) -> StreamExecutorResult<Vec<(GapFillCacheKey, CompactedRow, RowType)>> {
+        let end_time_row = self.time_key_serde.deserialize(end_time)?;
+
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) =
+            &(Bound::Unbounded, Bound::Excluded(end_time_row));
+
+        let state_table_iter = self
+            .state_table
+            .rev_iter_with_prefix(
+                &None::<row::Empty>,
+                sub_range,
+                PrefetchOptions::prefetch_for_large_range_scan(),
+            )
+            .await?;
+        pin_mut!(state_table_iter);
+
+        let mut results = Vec::new();
+
+        // Collect at most `limit` rows in reverse order.
+        while let Some(item) = state_table_iter.next().await {
+            let state_row = item?.into_owned_row();
+            let gapfill_row = self.get_gapfill_row(state_row);
+            results.push((
+                gapfill_row.cache_key,
+                (&gapfill_row.row).into(),
+                gapfill_row.row_type,
+            ));
+            if results.len() >= limit {
+                break;
+            }
+        }
+
+        results.reverse();
+
+        Ok(results)
+    }
+
+    /// Scans a range of data from `StateStore` to build a window of rows after `start_time`.
+    pub async fn scan_range_after(
+        &self,
+        start_time: &GapFillCacheKey,
+        limit: usize,
+    ) -> StreamExecutorResult<Vec<(GapFillCacheKey, CompactedRow, RowType)>> {
+        let start_time_row = self.time_key_serde.deserialize(start_time)?;
+
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) =
+            &(Bound::Excluded(start_time_row), Bound::Unbounded);
+
+        let state_table_iter = self
+            .state_table
+            .iter_with_prefix(
+                &None::<row::Empty>,
+                sub_range,
+                PrefetchOptions::prefetch_for_large_range_scan(),
+            )
+            .await?;
+        pin_mut!(state_table_iter);
+
+        let mut results = Vec::new();
+        let mut count = 0;
+
+        while let Some(item) = state_table_iter.next().await
+            && count < limit
+        {
+            let state_row = item?.into_owned_row();
+            let gapfill_row = self.get_gapfill_row(state_row);
+            results.push((
+                gapfill_row.cache_key,
+                (&gapfill_row.row).into(),
+                gapfill_row.row_type,
+            ));
+            count += 1;
+        }
+
+        Ok(results)
+    }
+
+    /// Converts a state table row to a `GapFillStateRow`.
+    fn get_gapfill_row(&self, state_row: OwnedRow) -> GapFillStateRow {
+        let is_filled = self.extract_is_filled_flag(&state_row);
```

Comment:
Let's add a new `extract_is_filled` which directly converts the datum to `RowType`.